rust/src/delta.rs

```diff
@@ -426,6 +426,49 @@ impl DeltaTableState {
     pub fn current_metadata(&self) -> Option<&DeltaTableMetaData> {
         self.current_metadata.as_ref()
     }
+
+    /// merges new state information into our state
+    pub fn merge(&mut self, new_state: &mut DeltaTableState) {
+        // build a lookup table of all remove actions
+
+        if !new_state.tombstones.is_empty() {
+            let new_removals: HashSet<&str> = new_state
+                .tombstones
+                .iter()
+                .map(|s| s.path.as_str())
+                .collect();
+
+            // previous state files must filtered for removals,
+            // new state files are appended without looking.
+            self.files
+                .retain(|a| !new_removals.contains(a.path.as_str()));
+        }
+        self.files.append(&mut new_state.files);
+        self.tombstones.append(&mut new_state.tombstones);
+        if new_state.min_reader_version > 0 {
+            self.min_reader_version = new_state.min_reader_version;
+            self.min_writer_version = new_state.min_writer_version;
+        }
+
+        if new_state.current_metadata.is_some() {
+            self.tombstone_retention_millis = new_state.tombstone_retention_millis;
+            self.current_metadata = new_state.current_metadata.take();
+        }
+
+        new_state
+            .app_transaction_version
+            .drain()
```

Comment:
if this method drains and mutates new_state, this means new_state is not usable after the call right? if so, I think it's better to pass in the new_state as value instead of reference to take full ownership.

Reply:
that sounds reasonable, I'll have a look :+1: